mitmproxy/tools/web/app.py

```diff
@@ -324,6 +327,72 @@ def broadcast(cls, **kwargs):
 class ClientConnection(WebSocketEventBroadcaster):
     connections: ClassVar[set] = set()
 
+    def __init__(self, application: Application, request, **kwargs):
+        super().__init__(application, request, **kwargs)
+        self.application: Application = application
+        self.filters: dict[str, flowfilter.TFilter] = {}  # Instance-level filters
+
+    def send_flow(self, cmd: str, f: mitmproxy.flow.Flow):
+        matches = [
+            {expr.pattern: bool(flowfilter.match(expr, f))}
+            for expr in self.filters.values()
+        ]
+        message = json.dumps(
+            {
+                "resource": "flows",
+                "cmd": cmd,
+                "matches": matches,
+                "data": flow_to_json(f),
+            },
+        ).encode("utf8", "surrogateescape")
+
+        self.send(conn=self, message=message)
+
+    def get_matching_flow_ids(self, name: str) -> List[str]:
+        expr = self.filters.get(name)
+        if not expr:
+            return []
+
+        return [f.id for f in self.application.master.view if expr(f)]
+
+    def send_matching_flow_ids(self, name: str, expr: str):
+        matching_flow_ids = self.get_matching_flow_ids(name)
+
+        message = json.dumps(
+            {
+                "resource": "flows",
+                "cmd": "filtersUpdated",
+                "name": name,
+                "expr": expr,
+                "data": matching_flow_ids,
+            },
+        ).encode("utf8", "surrogateescape")
```

Comment:
```suggestion
        ).encode()
```

We should never have surrogate escapes here, so let's keep it simple.